ls1intum/Artemis
# `Plagiarism checks`: Prevent multiple simultaneous plagiarism checks in the same course

### Checklist
#### General


#### Server
#### Client

### Motivation and Context



@krusche asked me to disable simultaneous plagiarism checks per course to avoid performance issues and to avoid that users execute multiple plagiarism checks accidentally in case they are slow (e.g. in exercises with 2000 participating users) 

### Description


To prevent simultaneous plagiarism checks per course, we store a set of course ids that indicates whether a plagiarism check is currently active. This map is stored in our distributed Hazelcast cache and therefore available on all instances.

If the user tries to run multiple plagiarism checks at the same time, an exception is thrown.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Exercises of one Course

1. Log in to Artemis
2. Navigate to Course Administration
3. Navigate to each exercise and the plagiarism check overview
4. Click both buttons of the plagiarism check at the same time
5. The second one should fail with an alert


### Review Progress





#### Code Review
#### Manual Tests







### Screenshots




https://user-images.githubusercontent.com/60005702/175806871-d4261fea-a1e8-407d-b918-56367e7c9080.mp4
